### ISSUE TITLE
reported "roll" for roll options is different from the acar.att.roll

### ISSUE THREAD

**Issue by jskrist** [sot/sparkles#95]
I might not fully understand this behavior.  This is what I am seeing.  When an html report is generated for an observation, if it has roll_options, a separate roll_options report page is generated.  The rolls reported on this page are different from what I am reporting in ORViewer.  Upon investigation of this discrepancy, I found that:

`acar.roll_options[1]["acar"].att.roll ~= acar.roll_options[1]["roll"]`

This seems to be caused by `acar.roll_options[1]["roll"]` being calculated using `(interval['x_start'] + interval['x_stop']) / 2`.  Is this expected?  Which is the "correct" roll?

**Comment by jeanconn** [comment]
It looks to me like acar.roll_options[1]["acar"].att.roll is ended up with roll in the ACA frame and acar.roll_optoins[1]["roll"] is ending up with a roll that is more in the target frame, but I could probably use a diagram.  @aldcroft did you intend all "saved" rolls to be back in the ACA frame?

**Comment by jeanconn** [comment]
My comment is made after setting a poor-man's breakpoint (unhanled exception) in the "if improvement > 0.3" block in get_roll_options and just looking at this for obsid 19409:

```
In [2]: aca = get_aca_catalog(19409)
In [4]: acar = aca.get_review_table()
In [5]: acar.get_roll_options()
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-85aa8e4b985b> in <module>()
----> 1 acar.get_roll_options()

/proj/sot/ska/jeanproj/git/aca_preview/sparkles/roll_optimize.py in get_roll_options(self)
    335                 acar = self.__class__(aca_rolled, obsid=self.obsid,
    336                                       is_roll_option=True)
--> 337                 raise ValueError
    338                 # Do the review and set up messages attribute
    339                 acar.check_catalog()

ValueError: 

In [9]: %debug
> /proj/sot/ska/jeanproj/git/aca_preview/sparkles/roll_optimize.py(337)get_roll_options()
    335                 acar = self.__class__(aca_rolled, obsid=self.obsid,
    336                                       is_roll_option=True)
--> 337                 raise ValueError
    338                 # Do the review and set up messages attribute
    339                 acar.check_catalog()

ipdb> acar.att.roll
191.11568595186216
ipdb> q_att_rolled.roll
191.11568595186216
ipdb> roll_interval['roll']
191.12098700000001
ipdb> q_targ_rolled.roll
191.12098700000001
```

**Comment by jeanconn** [comment]
I assume that you were seeing this for an OR and not an ER @jskrist ?

**Comment by jskrist** [comment]
@jeanconn - I was noticing this issue with ORs, but that was because I didn't have any ERs in the schedule with warnings.  I opened up another schedule and was able to verify that there is a discrepancy between the "roll" key in the roll_options dictionary, and the "roll" property of the same ACAResviewTable's "att" property.

**Comment by jskrist** [comment]
my understanding is that the mission planners (MPs) simply want to know what roll to provide to achieve the predicted result (e.g. no Critical messages).  I can report either of these values to the MPs (although when there are no roll_options calculated, I can only provide the acar.att.roll, which is why I am reporting that for all things at the moment).

There is also some confusion about what coordinate system the roll is provided in.  Honestly, I don't know what system is assumed in ORViewer, but ideally this would be transparent to the MPs.

**Comment by jeanconn** [comment]
Regarding "I opened up another schedule and was able to verify that there is a discrepancy between the "roll" key in the roll_options dictionary, and the "roll" property of the same ACAResviewTable's "att" property for both ORs and ERs." is that with commit
https://github.com/sot/sparkles/pull/93/commits/ad8c8200dafec1a9936420b9d72078fc7fe79363 ?

**Comment by jeanconn** [comment]
Regardin "Honestly, I don't know what system is assumed in ORViewer, but ideally this would be transparent to the MPs", I think they are used to dealing with ACA roll for ERs and target roll for ORs, but when we talked about this at the last meeting, Tom was suggesting that the rolls would likely all be in ACA frame and the mission planners could just fiddle around with them to try to get the catalog from the roll suggestions.  I don't know if other side-channel discussion has happened with that since.

**Comment by taldcroft** [comment]
Do you have handy a typical roll discrepancy?

**Comment by jskrist** [comment]
@jeanconn - I last pulled 4.1.1 this morning, so I haven’t updated since then. Not sure if ad8c820 is in my local code at the moment (on the mobile site away from my computer).

@taldcroft - typical difference is ~0.00#### where #### is anything from 0100 to 9999 (a rough estimate of the average is something like 0.0035-ish)

**Comment by taldcroft** [comment]
Sorry, I should not have bothered to ask that question, just thinking a little more would have saved you both time.  The roll discrepancy that one sees is a strong function of `dec`, and it basically goes singular at the poles:
```
In [6]: qt = calc_aca_from_targ([0, 89.9, 0], 0, 0)
In [7]: qt.roll0
Out[7]: 9.488776999148234
```
Just for some intuition, for a roll difference of less than about 100 arcsec, there is basically no difference from the ACA perspective because the stars will move by less than a pixel anywhere on the CCD (roughly speaking).  So if the target roll and ACA roll are within 100 arcsec (0.028 deg), we don't care.  The threshold is around dec = +/- 55 deg. 

As @jeanconn mentioned, last time we discussed this in the MATLAB tools WG, John was aware of potential inconsistency in the rolls.  But I'm happy enough to clean this up.  I would suggest that the `att` attribute is always ACA and a new attribute `att_targ` is always in "target" coordinates, which for ERs is the same as `att`.  So this means:
- Fix sparkles to ensure `att` is in ACA
- Add a `att_targ` property to the `BaseCatalogTable` class in proseco to provide the target attitude if requested.

**Comment by taldcroft** [comment]
FYI, I looked at the sparkles code and ensuring that everything is in the right frame and consistent is definitely do-able, but will take a little time for me to get it right.  I.e. I should be able to do this today.  In the meantime we should get the rest of the PRs merged.

**Comment by jeanconn** [comment]
And for @jskrist , my point about https://github.com/sot/sparkles/pull/93/commits/ad8c8200dafec1a9936420b9d72078fc7fe79363 was just that my understanding of that impact of that fix is that it changed things so that at least for ERs you should be seeing no difference.  Thanks!

**Comment by taldcroft** [comment]
Honestly I'm not sure if the code will work correctly right now at a dec of 89.9.  I had an idea to basically use the target attitude from the start of the `get_roll_options` method, and this is guaranteed to do everything in a consistent manner and make the internals simpler as well.  So it probably will be a straightforward fix.

**Comment by jeanconn** [comment]
Regarding Tom's plan above to scrub sparkes to ensure `att` is in ACA and to add an `att_targ` to `BaseCatalogTable`, I don't think that is strictly required for this release.

Right now, I think the labeling is a little confusing, but that the ORViewer interface should just use the "roll" or roll_interval['roll'] label in the lines of the per-observation reporting without more sparkles code changes.  If I'm understanding correctly that "roll" / roll_interval["roll"] value ends up being the roll that mission planners will want to plug in to ORViewer anyway (target roll for ORs, ACA roll for ERs).  

It would be my preference of course to have a separate attribute capture target attitude (which would have been great for testing and you know how much I love labels), but I'm not sure if there's a good use case.

However, I also thought that `att` was already in ACA frame everywhere (public), so if I am confused about the scope of this that's a separate issue.

**Comment by jeanconn** [comment]
Oops, and I was basically putting my comment in at the same time as https://github.com/sot/sparkles/issues/95#issuecomment-468284746 .  If you think that things could blow up as-is *and* think this is a straightforward fix that is good to get in, that's fine too.

**Comment by taldcroft** [comment]
> If I'm understanding correctly that "roll" / roll_interval["roll"] value ends up being the roll that mission planners want.

I am just a little worried about the handling of roll offsets vs. rolls and being sure what is what, and as mentioned above I would not be surprised now if it does something wrong at very high dec (e.g. when target roll is different by 5 deg from roll, and maybe target roll is 364 and ACA roll is 2).  I guess basically if we are OK with alpha-quality for this release (for roll options), then it's GO with the current set of fixes.  If we want to push back another day, we can get to beta.

As the validation notebook showed, the roll option stuff does not actually come into play that often now.

**Comment by jeanconn** [comment]
I think we can just hope for no observations of Polaris.

**Comment by taldcroft** [comment]
And BTW, I think for this go-round we could just put `att_targ` into `ACAReviewTable` and avoid a new proseco release.  Then proseco 4.5 will get that new property right away.

**Comment by jskrist** [comment]
@jeanconn, @taldcroft  - I pulled the master branch an hour ago, and I am still seeing this discrepancy (for obsid 21867.0):

![image](https://user-images.githubusercontent.com/991715/53587525-9a18fb00-3b58-11e9-9736-d3207a642148.png)

This should give a better idea of where this issue is showing up, and the kind of discrepancy I am seeing.

**Comment by jskrist** [comment]
sorry, I thought pr #97 was already merged into master.  I will get the PR and try it again.

**Comment by jeanconn** [comment]
I think #97 is not going to change what you are seeing; it adds an attribute to explicitly track the target roll and it makes the code easier to read regarding this issue.

**Comment by jeanconn** [comment]
We could add target RA Dec Roll to the text of the report for ORs if that would be less (or more) confusing.

**Comment by jskrist** [comment]
Mostly, I want to make sure I am reporting the "right" thing to the mission planners.  If they look at the ORViewer dialog and see a roll of 123.24**7**, but then look at the html report and see 123.24**2**, I know there are going to be questions about why these are different, and which one should be used/chosen.

**Comment by jeanconn** [comment]
I think they are actually used to this (pasting in one roll into ORViewer and having something else show up in SAUSAGE).  I think what you have is the right thing, but not sure if there is value added in also tracking the target attitude in the sparkles report.

**Comment by jeanconn** [comment]
Maybe the best thing is to just change the table header in the roll options view in your report to whatever the mission planners call this?  Replace "Roll (deg)" with "SI / Target / ORViewer  Roll" or whatever they call it to distiguish it from ACA frame roll?

**Comment by jskrist** [comment]
I know it has been suggested that the difference between these values is due to a difference in reference frames (ACA frame vs. SI frame I think).  However, looking at the code, it is not clear to me that this is the cause of the difference.  Is it true that the `acar.roll_options[n]["roll"]` value is in one of these frames and the `acar.roll_options[n]["acar"].att.roll` is in a different reference frame?

If this is the case, are ERs and ORs being treated the same?

Finally, we need a name/label for these two "rolls" and reference frames.  typing it out long-form each time is tiresome 😪 😜

**Comment by jeanconn** [comment]
> Is it true that the acar.roll_options[n]["roll"] value is in one of these frames and the acar.roll_options[n]["acar"].att.roll is in a different reference frame?

That's my inference from the code and is what @taldcroft indended (judging from the comments at the top of #97)  Whether it is true or not is a different question.  I put target-attitude-printing code for ORs in #103 if that is reassuring in any way.  I don't feel strongly that #103 is code that should go in the release.

> If this is the case, are ERs and ORs being treated the same?

"Same" is debatable in that context.  For ERs, target roll is the same as ACA roll.  In that regard ORs and ERs are handled differently.  I think that the roll in the the roll_options table should always be what mission planners use in ORViewer when they try to edit the attitude.  In that regard they are the same.  

> Finally, we need a name/label for these two "rolls" and reference frames. typing it out long-form each time is tiresome ￼ ￼

I agree!  My push back there was to see if the mission planners already had a preferred name/label before we picked something.

**Comment by taldcroft** [comment]
At least for now on nomenclature, I suggest we go with `ACA` and `target`.  In the sparkles code I consistently use `att` for the ACA frame and `att_targ` for the attitude in the respective frames.  In general I am using the `_targ` suffix to denote target frame, while the default (with no suffix) is ACA (because of proseco heritage where ACA is really all that is needed).

**Comment by taldcroft** [comment]
BTW, I pushed another commit to #97 that fixes a mistake in my first commit from this morning.  I said I would need the whole day to get this right... :smile:.

**Comment by jskrist** [comment]
That nomenclature makes sense, although I am not sure what that means for `acar.roll_options[n]["roll"]`, which doesn't directly have `att` or `att_targ` in its name.

**Comment by taldcroft** [comment]
So here is where Jean and I are on the same page that "roll" is "the roll that mission planners care about".  To be more clear for us and the code / API, in all the roll options code and outputs (after #97), "roll" and "roll_min" or "roll_nom" etc are **all in the target frame**, always.  With an `acar` object, there is no ambiguity as you are using either `att` or `att_targ`.

There is a caveat of course, namely that for an ER, the "target" frame is exactly the same as the ACA frame.  Mission planners are familiar and comfortable with this convention.  I'm not sure if they would use that wording (target frame for an ER), but I believe if you tell them that the roll values we are providing are always the one that they need, that will make them happy.  That's what John Scott was asking for previously.

I think a version of #103 is a good plan, where I would be even more explicit and use our nomenclature of `ACA` and `Target`.

**Comment by jeanconn** [comment]
My only go back wrt to using `Target` is that it now looks like we're saying the target is at the given coordinates (and all we are saying is the Target coordinates are in the Target reference frame).
